File: 251004/Krotsyuk/module-publisher/src/main/java/org/ikrotsyuk/bsuir/modulepublisher/service/WriterService.java
package org.ikrotsyuk.bsuir.modulepublisher.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import lombok.RequiredArgsConstructor;
import org.ikrotsyuk.bsuir.modulepublisher.dto.request.WriterRequestDTO;
import org.ikrotsyuk.bsuir.modulepublisher.dto.response.WriterResponseDTO;
import org.ikrotsyuk.bsuir.modulepublisher.entity.WriterEntity;
import org.ikrotsyuk.bsuir.modulepublisher.exception.JSONConverter;
import org.ikrotsyuk.bsuir.modulepublisher.exception.exceptions.NotFoundException;
import org.ikrotsyuk.bsuir.modulepublisher.exception.exceptions.UserWithSameLoginFoundException;
import org.ikrotsyuk.bsuir.modulepublisher.mapper.WriterMapper;
import org.ikrotsyuk.bsuir.modulepublisher.repository.WriterRepository;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

@RequiredArgsConstructor
@Service
public class WriterService {
    private final WriterRepository writerRepository;
    private final WriterMapper writerMapper;
    private final JSONConverter jsonConverter;

    @Cacheable(value = "writers")
    @Transactional(readOnly = true)
    public List<WriterResponseDTO> getWriters(){
        List<WriterEntity> writerEntityList = writerRepository.findAll();
        if(writerEntityList.isEmpty())
            return Collections.emptyList();
        else
            return writerMapper.toDTOList(writerEntityList);
    }

    @Cacheable(value = "writers", key = "#id")
    @Transactional(readOnly = true)
    public WriterResponseDTO getWriterById(long id){
        Optional<WriterEntity> writerEntityOptional = writerRepository.findById(id);
        return writerEntityOptional.map(writerMapper::toDTO).orElse(null);
    }

    @CacheEvict(value = "writers", allEntries = true)
    @Transactional
    public WriterResponseDTO addWriter(WriterRequestDTO writerRequestDTO) {
        Optional<WriterEntity> optionalWriterEntity = writerRepository.findByLogin(writerRequestDTO.getLogin());
        if(optionalWriterEntity.isEmpty())
            return writerMapper.toDTO(writerRepository.save(writerMapper.toEntity(writerRequestDTO)));
        else {
            try{
                throw new UserWithSameLoginFoundException(jsonConverter.convertObjectToJSON(writerRequestDTO));
            }catch (JsonProcessingException ex){
                return writerMapper.toDTO(optionalWriterEntity.get());
            }
        }
    }

    @CacheEvict(value = "writers", key = "#id")
    @Transactional
    public WriterResponseDTO deleteWriter(Long id){
        Optional<WriterEntity> optionalWriterEntity = writerRepository.findById(id);
        if(optionalWriterEntity.isPresent()) {
            writerRepository.deleteById(id);
            return writerMapper.toDTO(optionalWriterEntity.get());
        }else{
            throw new NotFoundException();
        }
    }

    @CacheEvict(value = "writers", allEntries = true)
    @Transactional
    public WriterResponseDTO updateWriter(Long id, WriterRequestDTO writerRequestDTO){
        Optional<WriterEntity> optionalWriterEntity = writerRepository.findById(id);
        if(optionalWriterEntity.isPresent()){
            WriterEntity writerEntity = optionalWriterEntity.get();
            writerEntity.setLogin(writerRequestDTO.getLogin());
            writerEntity.setPassword(writerRequestDTO.getPassword());
            writerEntity.setFirstname(writerRequestDTO.getFirstname());
            writerEntity.setLastname(writerRequestDTO.getLastname());
            return writerMapper.toDTO(writerRepository.save(writerEntity));
        }else
            throw new NotFoundException();
    }
}

File: 251003/Pankratiev/rest/publisher/src/main/kotlin/bsuir/dc/publisher/service/WriterService.kt
package bsuir.dc.publisher.service

import bsuir.dc.publisher.dto.from.WriterRequestTo
import bsuir.dc.publisher.dto.to.WriterResponseTo
import bsuir.dc.publisher.mapper.toEntity
import bsuir.dc.publisher.mapper.toResponse
import bsuir.dc.publisher.repository.IssueRepository
import bsuir.dc.publisher.repository.WriterRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.stereotype.Service

@Service
class WriterService(
    private val writerRepository: WriterRepository,
    private val issueRepository: IssueRepository,
) {
    @Caching(
        evict = [
            CacheEvict(value = ["writers"], key = "'all_writers'")
        ]
    )
    fun createWriter(writerRequestTo: WriterRequestTo): WriterResponseTo {
        val writer = writerRequestTo.toEntity()
        val savedWriter = writerRepository.save(writer)
        return savedWriter.toResponse()
    }

    @Cacheable(value = ["writers"], key = "#id")
    fun getWriterById(id: Long): WriterResponseTo {
        val writer = writerRepository.findById(id).orElseThrow { NoSuchElementException() }
        return writer.toResponse()
    }

    @Cacheable(value = ["writers"], key = "'all_writers'")
    fun getAllWriters(): List<WriterResponseTo> =
        writerRepository.findAll().map { it.toResponse() }

    @Caching(
        put = [CachePut(value = ["writers"], key = "#id")],
        evict = [CacheEvict(value = ["writers"], key = "'all_writers'")]
    )
    fun updateWriter(id: Long, writerRequestTo: WriterRequestTo): WriterResponseTo {
        val updatedWriter = writerRequestTo.toEntity().apply { this.id = id }
        return writerRepository.save(updatedWriter).toResponse()
    }

    @Caching(
        evict = [
            CacheEvict(value = ["writers"], key = "#id"),
            CacheEvict(value = ["writers"], key = "'all_writers'")
        ]
    )
    fun deleteWriter(id: Long) {
        writerRepository.findById(id).orElseThrow { NoSuchElementException() }
        writerRepository.deleteById(id)
    }

    @Cacheable(value = ["writersByIssue"], key = "#issueId")
    fun getWriterByIssueId(issueId: Long): WriterResponseTo? {
        val issue = issueRepository.findById(issueId).orElseThrow { NoSuchElementException() }
        return issue.writer.toResponse()
    }
}